### ISSUE TITLE
Problem of reproducing the quality of the speech from "web_UI"

### ISSUE THREAD

**Issue by frank613** [e-c-k-e-r/vall-e#12]
Hi Ecker,

I managed to generate very decent TTS samples from your web-UI (https://huggingface.co/spaces/ecker/vall-e). I thought at first that web was running based on the v2-reference-models(nemo-smaller/larger-44khz-llama-8). However I failed to generate comparable samples locally with both v2 models. I tried many different combinations of sampling parameters but still the generated speech is either silence, low volume, or incomplete.  Now I doubted that it might actually the v1-model is running behind and with which I could generate good samples locally with similar scripts.  Could you please inform which is the model used for the web-UI?

If the actual model for the web-UI is v1, then I believe I must made some mistakes in the script which I would like to paste here. The function that I used for generating the speech  is an analog  to the part in "inference.py" file.
`def get_tts_results(model, text_in, prop_in, lang, device, reps_in, score_masked_only=False, out_path=None): 
    #len
    input_kwargs = dict(
                phns_list=[text_in], 
                raw_text_list=None,
                proms_list=[prop_in],
                lang_list=[lang],
            )
    len_list = model( **input_kwargs, task_list=["len"])
    #tts
    input_kwargs = dict(
                phns_list=[text_in], 
                proms_list=[prop_in],
                lang_list=[lang],
                task_list=["tts"],
                len_list=len_list,
                disable_tqdm=False,
                use_lora=True,
            )

    for i in range(3):
        sampling_kwargs = {"temperature": 1, "cfg_strength": 3, "max_steps":50}
        resps_list_out = model( **input_kwargs, **sampling_kwargs)
        ## decode
        Path(out_path).mkdir(parents=True, exist_ok=True)
        for idx, reps_out in enumerate(resps_list_out):
            wav, sr = qnt.decode_to_file(reps_out, out_path+f"test-full-{i}.wav", device=device) 
  
    _logger.info(f"decoding done")`

I checked the input to those models and they are really in a good shape (at least work for V1 model). Please point out if you could spot anything goes wrong here for V2.

Thanks,
Xinwei

**Comment by e-c-k-e-r** [comment]
> I managed to generate very decent TTS samples from your web-UI (https://huggingface.co/spaces/ecker/vall-e)

Huh, I could have sworn the HF space was in a degraded state. Last I checked there was some error about the ZeroGPU runtime or some such. I guess it fixed itself.

> I thought at first that web was running based on the v2-reference-models(nemo-smaller/larger-44khz-llama-8
> Could you please inform which is the model used for the web-UI?

By default, if no model is requested, it'll download and load the "reference" `ar+nar-len-llama-8` 24KHz EnCodec-based model. I think the simplest solution is for me to spin a second space that defaults to the `nvidia/audio-codec-44khz`-branded model, since HuggingFace spaces are a bit agonizing to cooperate.

> However I failed to generate comparable samples locally with both v2 models. I tried many different combinations of sampling parameters but still the generated speech is either silence, low volume, or incomplete.

This is something that I can't really elucidate as I can't recall much on PyTorch-setup-specific problems beyond it seeming to be tied to the specific PyTorch/ROCm version I had set up. I think it was some flavor of stable `2.6.x` that had a broken attention mechanism(s), and `flash_(sdpa)` was the only one that "worked". If I recall right, the nightly version at the time had a different set of problems. Setting up a fresh venv for PyTorch/ROCm `2.7.0` (stable) seemed to magically fix whatever issues I remember having.
* I can't recall if the typical PyTorch/CUDA releases had any quirks but I wouldn't be surprised if I did have to chase some gremlins pertaining to it.

However, I think the latest weights for the `larger` variant are inherently broken from doing some experimental training that I think I erroneously uploaded. [https://huggingface.co/ecker/vall-e/blob/944c5b33fbe367be392b5a0ffbb409ddb457b5cb/models/ckpt/nemo-larger-44khz-llama-8/fp32.sft](The prior commit) I think worked better in my brief LoRA testing last night, but the revision prior to that might be better.

The `smaller` variant is probably botched since it was erroneously trained under the afformentioned PyTorch/ROCm with the bugged attention mechanism. When I get the time and mental faculties I'll evaluate it and see about "fixing" it.

My "good-enough" settings yesterday seemed to be just loading the model under `bfloat16` with `sdpa` attention. Sampler settings to their defaults in the web UI, but sometimes checking `masked tokens only` + `remask` (or whatever I called them).

> Please point out if you could spot anything goes wrong here for V2.

At a glance it seems right. The "interface" for the v1/v2 models shouldn't change.

I would say that the only problem(s) would be:
* ensuring your inputs passed through are properly tokenized
* for loading the v2, ensure the model is loaded through `--model="./path/to/the/fp32.sft"`.

Whenever I get a chance I can see about validating the model works when interfaced per your script. I typically stick with using the `TTS` class exposed through `vall_e.inference` since it handles the finer details.

**Comment by frank613** [comment]
Hi, 

I think I got better results now from local with checked "masked tokens only + remask". Thanks a lot for the suggestions. 

But still the quality is not comparable to the v1 model. That's fine because there is a always a space to improve 👍 .  I am actually using your model to do some speech assessment research. The idea of FSQ is quite attractive to me for my research because of the way they split the mel-spectrogram.  I will keep watching this great work! 

Some notes:

> bfloat16 with sdpa attention

I load the config from the reference model "fp32.sft", and I checked indeed cfg.weights_name="fp32", do you mean mean one needs to do a weights conversion or something? 

> ensuring your inputs passed through are properly tokenized

Actually I ran into a trap a few weeks ago when I used all lowercase text input. The reason is it seems that for lowercase input (no punctuations) the tokeniser will generate no "<UNK>" according to its vocabulary and as a result the "g2p" part is skipped. It is the same case with the web-UI as well. I think it worths giving a warning or something.  

```
def encode_text(text, language="auto", precheck=True, phonemize=True ):
    # already a tensor, return it
    if isinstance( text, torch.Tensor ):
        return text
    # check if tokenizes without any unks (for example, if already phonemized text is passes)
    if precheck and "<unk>" in phn_symmap:
        tokens = tokenize( text )
        if phn_symmap["<unk>"] not in tokens:
            return torch.tensor( tokens )

    if not phonemize:
        return torch.tensor( text_tokenize( text ) )

    return torch.tensor( tokenize( g2p.encode(text, language=language) ) )
```

Thanks!!
